server/aesdsocket.c

```diff
@@ -14,140 +14,288 @@
 #include<string.h>
 #include<arpa/inet.h>
 #include<netinet/in.h>
+#include "queue.h"
+#include <pthread.h>
+#include <time.h>
+#include <sys/time.h>
+#include <string.h>
 
-// struct sockaddr{
-//     sa_family_t sa_family;
-//     char sa_data[14];
-// }
+//pthread_mutex_t *mutex;
 
 #define PORT "9000"
-#define BACKLOG (6)
+#define BACKLOG (5)
 #define BUFFER_SIZE (1024)
 #define FILE_PATH "/var/tmp/aesdsocketdata"
 
+pthread_mutex_t mutex;
+int sockfd,new_sockfd,file_fd;
+bool interrupted=false;
+char* buf=NULL;
+struct addrinfo *server_info=NULL;
+int total_connections=0;
+bool timer_alarm=false;
+
+struct thread_data{
+    pthread_t thread_id;
+    int socketfd;
+    struct sockaddr *client;
+    bool connection_complete_success;
+    SLIST_ENTRY(thread_data) entries;
+};
+
+SLIST_HEAD(slisthead,thread_data) head=SLIST_HEAD_INITIALIZER(&head);
+
+int set_time(){
+    int ret;
+    struct itimerval timer_1;
+    getitimer(ITIMER_REAL,&timer_1);
+    timer_1.it_value.tv_sec=10;
+    timer_1.it_value.tv_usec=0;
+    timer_1.it_interval.tv_sec=10;
+    timer_1.it_interval.tv_usec=0;
+    ret=setitimer(ITIMER_REAL,&timer_1,NULL);
+    return(ret);
+}
+int reset_time(){
+    int ret;
+    struct itimerval timer_1;
+    getitimer(ITIMER_REAL,&timer_1);
+    timer_1.it_value.tv_sec=0;
+    timer_1.it_value.tv_usec=0;
+    timer_1.it_interval.tv_sec=0;
+    timer_1.it_interval.tv_usec=0;
+    ret=setitimer(ITIMER_REAL,&timer_1,NULL);
+    return(ret);
+}
+
+int write_time_to_file(int fd){
+    char outstr[100]={};
+    char buffer[100];
+    time_t t;
+    struct tm *tmp;
+    const char* fmt= "%a, %d %b %Y %T %z";
+    t = time(NULL);
+    tmp = localtime(&t);
+    if (tmp == NULL) {
+        perror("localtime");
+        return(-1);
+    }
+
+    if (strftime(outstr, sizeof(outstr), fmt, tmp) == 0) {
+        return(-1);
+    }
+    strcpy(buffer,"timestamp:");
+    strcat(buffer,outstr);
+    strcat(buffer,"\n");
+    write(fd,buffer,strlen(buffer)*sizeof(char));
```

Comment:
@Sona-Subair see the assignment requirements regarding write to `fd` here and locking